### ISSUE TITLE
Libby setup codes don't work

### ISSUE THREAD

**Issue by ParsonsGlory** [ping/odmpy#78]
Neither of the remaining options for generating setup codes works. The code for sonos speakers and the code for android auto generates the error: Could not log in with code. 

Any suggestions anyone?

**Comment by coloradoruss** [comment]
When I set mine up a while ago the code was generated by the sending
account and the receiving account just entered the code and set up. This is
how the software worked not too long ago. They seem to have reversed the
codes now the receiving account generates a code and it tells you to enter
the code in the account that you want to copy from. I don't know if the
software has been modified to catch this.

**Comment by wtanksleyjr** [comment]
Oh no. I can give some more detail - the failure is that one of the Libby steps sends an invalid certificate, which Python's parser rejects. odmpy seems to expect the error and just reports that authentication didn't work, but more is going on than usual. This discussion should be brought under the old thread #69, since it seems to belong to the same category.

**Comment by ParsonsGlory** [comment]
The invalid certificate error seems to have been fixed https://github.com/ping/libby-calibre-plugin/issues/33. Hopefully this work will be migrated to odmpy,

Edit: Fixed here https://github.com/jordkh/odmpy

**Comment by wtanksleyjr** [comment]
FYI, the message posted 2 weeks ago was edited to include a solution - if you were watching this post for a solution you wouldn't see it because edits don't get sent out to watchers, so I'm reposting it.

Hat tip: @ParsonsGlory 
> Fixed here https://github.com/jordkh/odmpy

**Comment by coloradoruss** [comment]
Awesome.  Thank you!  Great to have that fixed.



Russ Breckenridge
Pink Realty Rentals
Property Management

Direct Contact:
(719) 630-1000 - Direct, Call only, Do not text
(303) 380-1000 - Direct, Call only, Do not text
(719) 629-7465 - Text message only
***@***.***

Office Contact:
(719) 393-7465 - Pink Realty, Main number
(719) 633-7000 - Property Management
(719) 260-6000 - Pink Realty Buyer Team
(719) 227-7000 - Pink Realty Seller Team
(719) 203-9000 - Transaction Coordinator
(800) 677-3253 - Fax

Check out our website for full access to all homes listed for sale in the
Pikes Peak, Denver, Pueblo, and Royal Gorge MLS systems.  www.PinkRealty.com
<http://www.pinkrealty.com/>

Check out our website for all of our currently available rental
properties.  www.PinkRentals.com



*Confidentiality Notice: This message may contain confidential and/or
privileged information.  If you are not the addressee or authorized to
receive this for the addressee, you must not use, copy, disclose, or take
any action based on this message or any information herein.  If you have
received this message in error, please advise the sender immediately by
reply e-mail and delete this message.Non-Binding and Expiration Disclaimer:
Content of emails, to include any included links and attachments thereto,
sent or received shall not create a binding contract or be construed to be
conveyance or acceptance of any offer, contract or any other agreement,
whatsoever, nor shall the same be binding on any party until a written
contract is signed by all parties.  In the event a bona fide, signed offer
is conveyed via email, the same shall expire, if not accepted and fully
executed by all parties, upon the latter of its explicitly stated
expiration date or, if such a date is not specified, twenty-four hours from
its initial transmission by the offeror and shall thereafter not be binding
on any party.*




*Wire Fraud Notice: Never trust wiring instructions sent via email. Cyber
criminals are hacking email accounts and sending emails with fake wiring
instructions. These emails are convincing and sophisticated. Always
independently confirm wiring instructions in person or via a telephone call
to a trusted and verified phone number. Never wire money without
double-checking that the wiring instructions are correct.Affiliated
Business Disclosure: I have an ownership interest in both Pink Realty, Inc
and Pink Realty Rentals and work closely with both companies and their
agents and employees, my real estate license and my brokerage activity is
with Pink Realty Rentals.*

**Comment by dideler** [comment]
Any ideas how to get past the first-run of `odmpy libby` since Libby reversed the setup code flow as mentioned in https://github.com/ping/odmpy/issues/78#issuecomment-2357461176?

The code is now given from the device that has to recover the data. In other words, you can no longer get a code from a logged in session of Libby to give to odmpy.

Edit: Using the locally installed fork https://github.com/jordkh/odmpy with the alternative Sonos flow worked as it still has the logged-in session generate a code.
<img width="473" alt="image" src="https://github.com/user-attachments/assets/bfc8d606-0605-43e6-8a30-f0d7b2bd3ad5">

Now running into another error getting a 403 when opening an audiobook URL. But was able to get around that by downloading the ODM file manually from Overdrive web as in https://github.com/ping/odmpy/issues/70#issuecomment-2081232634 and https://github.com/ping/odmpy/issues/69#issuecomment-2078355058.

**Comment by wtanksleyjr** [comment]
Yes, from where it asks you for the code (instead of showing the code) you
just click on either of the two options underneath it (Android Car, or
Sonos); either one will give you a code in the old way.

-Wm

**Comment by oandpg** [comment]
Not working for me using the sonos/android/libby codes. still getting the same error.

**Comment by wtanksleyjr** [comment]
That's because it's not fixed in this repository, but it has been fixed at https://github.com/jordkh/odmpy . 
The command I just ran to install this was `python3 -m pip install git+https://git@github.com/jordkh/odmpy --upgrade --force-reinstall` .

**Comment by oandpg** [comment]
Yes I installed it from the jordkh repo but it had not installed properly. can now get it to accept the code, and it displays the books, but any attempt to download results in a 403 error. The suggested workarounds of using an ODM file will disappear on the 13 nov when they stop displaying the mp3 option...

**Comment by wtanksleyjr** [comment]
Yes, they effectively shut Libby down for us a long time ago; the ODM
workaround was all we had.

You're right that's going away next month.

And that's it for us.

I'm told that people have gotten mp3 files downloaded from the Libby
interface, but this is not a part of programming I've dabbled in, and ping
(the original author of odmpy) has been gone for a while.

-Wm

**Comment by jahlborn** [comment]
now that odm files can't be downloaded anymore, does anyone have any working solutions at this point?

**Comment by wtanksleyjr** [comment]
Not really. Sorry, I wish we had the expertise, but with the author MIA and no other expert at hand there's not much to be done... I'm a programmer, but in this field I'm just flailing about.

**Comment by joshcsmith13** [comment]
I was able to use the forked version at https://github.com/jordkh/odmpy to complete the setup.  Then, in order to get past the error downloading, I used the code and instructions for this "bookmarklet" to expose the "Download MP3" link on the Overdrive Loans web page: https://gist.github.com/ping/b58ae66359691db1d08f929a9e57a03d
(I also had to find/install ffmpeg and add it to my Windows PATH.)
Once I had the ODM file, I could run 'odmpy dl "filename.odm"' to get the mp3s.  
whew...

**Comment by joshcsmith13** [comment]
You do have to run "odmpy dl" shortly after downloading the file or you will get an error about an expired odm file.

**Comment by jay3fer** [comment]
> I was able to use the forked version at https://github.com/jordkh/odmpy to complete the setup. Then, in order to get past the error downloading, I used the code and instructions for this "bookmarklet" to expose the "Download MP3" link on the overdrive/Libby web page: https://gist.github.com/ping/b58ae66359691db1d08f929a9e57a03d (I also had to find/install ffmpeg and add it to my Windows PATH.) Once I had the ODM file, I could run 'odmpy dl "filename.odm"' to get the mp3s. whew...

Amazing, thank you so much!!! I have my audiobooks back... :)

**Comment by wtanksleyjr** [comment]
Is anyone else getting nothing but errors from `odmpy dl` again? I can use the script to get odms, but the download seems to be blocked. At least for me. Anyone?

**Comment by t-shizzle** [comment]
> Is anyone else getting nothing but errors from `odmpy dl` again? I can use the script to get odms, but the download seems to be blocked. At least for me. Anyone?

I just ran another download and it worked great.  after getting the bookmark I run this at the command prompt
odmpy dl -d "c:\bot"  "C:\Users\X\Downloads\Booktitle.odm"

After getting teh bookmark I just highlight the file and mouse drag the name to the command prompt and it brings over the file address.  it all seems to work great for me.  Have you tried it again today?

**Comment by wtanksleyjr** [comment]
Thank you for checking! I just tried again ... and this time it worked! WHEW.

**Comment by wolfkin** [comment]
Got nothing yet.

1.  Used the following to setup from the fork
`python -m pip install git+https://git@github.com/jordkh/odmpy --upgrade --force-reinstall`

2. Completed setup `odmpy libby` using Sonos code.

3. Took bookmarklet from internal instructions [here](https://gist.githubusercontent.com/ping/b58ae66359691db1d08f929a9e57a03d/raw/03b945138147afa5d94a821c525ba5c1346d0b64/bookmarklet.js).

4. Opened up audiobook, clicked bookmarklet and nada. 

5. Attempted `odmpy libby` and I do get the interactive client that shows my listings.

```
odmpy Interactive Client for Libby
----------------------------------------------------------------------
Found 1 loan.
 1: Butter Honey Pig Bread                            Francesca Ekwuyasi
    * Expires: 2024-12-23  XXXXXXX Public Library

Download. Choose from 1-1 (separate choices with a space or leave blank to quit),
then press enter:
```
[trace1.txt](https://github.com/user-attachments/files/18172279/trace1.txt)

6. trying to download from the ODM isn't working either.
[trace2.txt](https://github.com/user-attachments/files/18172296/trace2.txt)

The ODM itself seems empty anyway
[ODM_Contents.txt](https://github.com/user-attachments/files/18172304/ODM_Contents.txt)



7. Just doubled checked and I do have `ffmpeg` installed and accessible from other things. Heck I still have Overdrive installed. We only got cut off OD a few weeks ago.

**Comment by wtanksleyjr** [comment]
You're good that the setup worked.

Instead of step 4, you should click the bookmarklet from your loans page, not from an "opened-up audiobook" (I'm not sure what that means BTW). In my case my loans page is at https://lapl.overdrive.com/account/loans. What it does is change the page so that all of the "Have Overdrive for Mac/Windows? >" links reveal their "Download" buttons (like they used to).

As step 5, you can then click on the Download links under the books you want, and you'll get the ODM file you'd have expected. Next you copy it to the machine running odmpy, and run `odmpy dl <whatever the filename is>`, with whatever options you want (I like `-k` to keep the cover.jpg for example).

**Comment by joshcsmith13** [comment]
you lost me with "#4: Opened up audiobook"  You want to click/open/run the bookmarklet while on the Overdrive.com LOANS page.  That should cause the "Download MP3 audiobook" link to appear. 
![image](https://github.com/user-attachments/assets/1ac8f638-757f-4066-88f6-c1c3e2f3aeb2)

....
what wtanksleyjr said!  :)


I just did it today, so I'm confident that it still works.
![image](https://github.com/user-attachments/assets/21b0d60d-43df-471d-9efe-068581c092b9)

**Comment by wolfkin** [comment]
Ahh I see. I was trying the bookmarklet on Libby but I should have been using it on the Library page. That DID work. thank you.

The ODM failed in Overdrive. But it does seem to be working with `odmpy`. Incredibly slow... but working. ~~I was going to wait until it finished for an update but I'll just edit this later and let this run for now.~~

Nevermind. Turns out it _was_ my VPN. Download failed just before I hit submit and when I flicked it off it was golden. Regular speeds.

![scz5xj08Db (Small)](https://github.com/user-attachments/assets/d5db3506-ac1f-4054-9844-d960319b1605)

My HatTip to you all. I didn't expect to get this all solved tonight.

**Comment by talosgl** [comment]
> Got nothing yet.
> 
> 1. Used the following to setup from the fork
>    `python -m pip install git+https://git@github.com/jordkh/odmpy --upgrade --force-reinstall`
> 2. Completed setup `odmpy libby` using Sonos code.

I'm having trouble getting past this step to authenticate with the Sonos or the Android Auto code. I *think* I have latest on the fork:
```$ odmpy --version```
```odmpy 0.8.1 [Python 3.13.1-win32]```

I get this error every time:
```Error: Could not log in with code.```
```Make sure that you have entered the right code and within the time limit.```

**EDIT/SOLVE**: Apparently I don't need this for ```odmpy dl <filepath.odm>``` to work, because that works fine. Leaving my comment here in case someone else hits the same issue! (You will need the [bookmarklet ](https://gist.github.com/ping/b58ae66359691db1d08f929a9e57a03d)already set up to get that file).

**Comment by wtanksleyjr** [comment]
> **EDIT/SOLVE**: Apparently I don't need this for `odmpy dl <filepath.odm>` to work, because that works fine. Leaving my comment here in case someone else hits the same issue! (You will need the [bookmarklet ](https://gist.github.com/ping/b58ae66359691db1d08f929a9e57a03d)already set up to get that file).

Yes, the fix discussed in this issue and available at the fork is ONLY needed if you want to do things like get a list of all the books you have checked out. It's not super-useful, although I have scripts that connect the two so that I can use another system to get book metadata.

It looks to me like you're running the version of odmpy at this site, not the one at the fork. Others have posted a Python command line which will install that one (and you might want to remove this one first).

**Comment by talosgl** [comment]
> It looks to me like you're running the version of odmpy at this site, not the one at the fork. Others have posted a Python command line which will install that one (and you might want to remove this one first).

You're completely right. What I goober I am. Thanks for taking the time to reply and help me learn!
